app/wasm/message_plugin.go

```diff
@@ -0,0 +1,201 @@
+package wasm
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+
+	wasmbindings "github.com/osmosis-labs/osmosis/v7/app/wasm/bindings"
+	gammkeeper "github.com/osmosis-labs/osmosis/v7/x/gamm/keeper"
+	gammtypes "github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+)
+
+func CustomMessageDecorator(gammKeeper *gammkeeper.Keeper, bank *bankkeeper.BaseKeeper) func(wasmkeeper.Messenger) wasmkeeper.Messenger {
+	return func(old wasmkeeper.Messenger) wasmkeeper.Messenger {
+		return &CustomMessenger{
+			wrapped:    old,
+			bank:       bank,
+			gammKeeper: gammKeeper,
+		}
+	}
+}
+
+type CustomMessenger struct {
+	wrapped    wasmkeeper.Messenger
+	bank       *bankkeeper.BaseKeeper
+	gammKeeper *gammkeeper.Keeper
+}
+
+var _ wasmkeeper.Messenger = (*CustomMessenger)(nil)
+
+func (m *CustomMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) ([]sdk.Event, [][]byte, error) {
+	if msg.Custom != nil {
+		// only handle the happy path where this is really minting / swapping ...
+		// leave everything else for the wrapped version
+		var contractMsg wasmbindings.OsmosisMsg
+		if err := json.Unmarshal(msg.Custom, &contractMsg); err != nil {
+			return nil, nil, sdkerrors.Wrap(err, "osmosis msg")
+		}
+		if contractMsg.MintTokens != nil {
```

Comment:
Is it worth adding any error handling if multiple fields are entered (e.g. no result if MintTokens != nil and Swap != nil)

Reply:
Thought about that, and decided not to, as it seemed overkill. We can add it, for correctness.

Reply:
It is impossible to do this in Rust if you use our bindings. Only if you make some custom types trying to break the caller.

If the worse we handle when you send invalid data is interpreting it as MintToken and ignoring the swap, I don't think this is an issue